File: c++/c06_1_coin_problem.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
constexpr ll INF = numeric_limits<ll>::max() / 2;

namespace {
	ll solve(vector<ll> coins, ll n) {
		if (n < 0)
			return INF;
		ll count = 0;
		ranges::sort(coins, greater());
		for (ll c : coins) {
			count += n / c;
			n %= c;
		}
		return (n == 0) ? count : INF;
	}
}

TEST(C061CoinProblem, solve) {
	vector<ll> coins = { 1,2,5,10,20,50,100,200 };
	EXPECT_EQ(solve({}, 0), 0);
	EXPECT_EQ(solve({}, 1), INF);
	EXPECT_EQ(solve({ 5 }, 3), INF);
	EXPECT_EQ(solve(coins, -1), INF);
	EXPECT_EQ(solve(coins, 0), 0);
	EXPECT_EQ(solve(coins, 520), 4);
	EXPECT_EQ(solve({ 200,1,2,5,10,20,50,100 }, 520), 4);
}

File: python/c06_1_coin_problem.py
import unittest

INF=1<<63

def solve(coins,n):
    if n<0:
        return INF
    count=0
    for c in sorted(coins,reverse=True):
        count+=n//c
        n%=c
    return count if n==0 else INF

class Test(unittest.TestCase):
    def test(self):
        coins=[1,2,5,10,20,50,100,200]
        self.assertEqual(solve([],0),0)
        self.assertEqual(solve([],1),INF)
        self.assertEqual(solve([5],3),INF)
        self.assertEqual(solve(coins,-1),INF)
        self.assertEqual(solve(coins,0),0)
        self.assertEqual(solve(coins,520),4)
        self.assertEqual(solve([200,1,2,5,10,20,50,100],520),4)

if __name__=='__main__':
    unittest.main()
